ompi/mca/coll/han/coll_han_subcomms.c

```diff
@@ -0,0 +1,449 @@
+/*
+ * Copyright (c) 2018-2020 The University of Tennessee and The University
+ *                         of Tennessee Research Foundation.  All rights
+ *                         reserved.
+ * Copyright (c) 2020      Bull S.A.S. All rights reserved.
+ *
+ * $COPYRIGHT$
+ *
+ * Additional copyrights may follow
+ *
+ * $HEADER$
+ */
+/**
+ * @file
+ *
+ * Warning: this is not for the faint of heart -- don't even bother
+ * reading this source code if you don't have a strong understanding
+ * of nested data structures and pointer math (remember that
+ * associativity and order of C operations is *critical* in terms of
+ * pointer math!).
+ */
+
+#include "ompi_config.h"
+
+#include "mpi.h"
+#include "coll_han.h"
+#include "coll_han_dynamic.h"
+
+
+/*
+ * Local functions
+ */
+static void create_intranode_comm_new(ompi_communicator_t *,
+                                      ompi_communicator_t **);
+static void create_internode_comm_new(ompi_communicator_t *,
+                                      int, int,
+                                      ompi_communicator_t **);
+static void create_intranode_comm(ompi_communicator_t *,
+                                  const char *,
+                                  int,
+                                  ompi_communicator_t **);
+static void create_internode_comm(ompi_communicator_t *,
+                                  const char *,
+                                  int, int,
+                                  ompi_communicator_t **);
+
+/**
+ * Create a sub-communicator containing the ranks that share my node.
+ *
+ * @param comm (IN)          original communicator for the collective
+ *                           target module priority
+ * @param sub_comm (OUT)     created sub-communicator
+ */
+static void create_intranode_comm_new(ompi_communicator_t *comm,
+                                  ompi_communicator_t **sub_comm)
+{
+    ompi_comm_split_type(comm, MPI_COMM_TYPE_SHARED, 0,
+                         (opal_info_t *)(&ompi_mpi_info_null), sub_comm);
+    return;
+}
+
+/**
+ * Create a sub-communicator containing one rank per node.
+ *
+ * @param comm (IN)          original communicator for the collective
+ * @param my_rank (IN)       my rank in comm
+ * @param intra_rank (IN)    local rank in the intra-node sub-communicator
+ * @param sub_comm (OUT)     created sub-communicator
+ */
+static void create_internode_comm_new(ompi_communicator_t *comm,
+                                  int my_rank,
+                                  int intra_rank,
+                                  ompi_communicator_t **sub_comm)
+{
+    ompi_comm_split(comm, intra_rank, my_rank, sub_comm, false);
```

Comment:
This function is incorrect, it does not only create a communicator between leaders (one process per node), it creates a communicator between similarly ranked processes per node. Which means that some of the communicators will not contain processes on all nodes (except if the original communicator contains exactly the same number of processes on each node which might be unlikely).

Moreover, we only need to create a communicator between rank 0 on each node, so the correct call should be
```ompi_comm_split(comm, 0 == intra_rank ? 0 : MPI_UNDEFINED, my_rank, sub_comm, false);```

Reply:
Indeed, we also noticed this issue in the original code ( https://github.com/bosilca/ompi/blob/coll/han/ompi/mca/coll/han/coll_han_module.c#L280 for example), but some algorithm take advantage of these multiple communicator per nodes to select the most performant one. For example a MPI_Gather with a root on local N (different from 0) will perform the internode gather using the Nth intercommunicator. This way a final copy is avoided. This may be interesting in cases of large amount of data, which seems to be one of the target for topological collectives.
As you mentionned when the number of process per node is imbalanced, the communicator may not cover all the nodes. In this case, it planned to temporarely fallback to another component (until algorithms with local copy are implemented)

Reply:
Recents commit should fix all functional issues in case of imbalanced process distribution : reduce, scatter, gather and bcast fallback on previous components. Other collectives always use leader rank  (local rank 0) for internode communications.